### PR TITLE
Use align_method in comp_method_FRAME

### LINKED ISSUE #20090: Mixed-dtype dataframe comparison with array raises incorrectly
This came up while going through some of statsmodels tests:

```
arr = np.random.randn(3, 2)
arr[:, 0] = [1, 2, 3]
df = pd.DataFrame(arr)
df[0] = df[0].astype(int)
>>> df == arr
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/ops.py", line 1572, in f
    try_cast=False)
  File "pandas/core/frame.py", line 4021, in _combine_const
    try_cast=try_cast)
  File "pandas/core/internals.py", line 3644, in eval
    return self.apply('eval', **kwargs)
  File "pandas/core/internals.py", line 3538, in apply
    applied = getattr(b, f)(**kwargs)
  File "pandas/core/internals.py", line 1348, in eval
    t_shape=values.T.shape, oth_shape=other.shape))
ValueError: cannot broadcast shape [(3, 1)] with block values [(3, 2)]
```

I'd expect this to wrap the ndarray in a frame and return an all-True frame.

### PR DESCRIPTION
**Pull request by jbrockmendel**
Closes #20090

<b>update</b> Since #23000 was merged, some of the discussion is out of date.  The bottom line remains unchanged: This PR makes DataFrame _comparison_ ops behave like DataFrame _arithmetic_ ops currently do.  Also fixes some bugs e.g. #20090
<b> end update</b>

This is a much nicer alternative to the implementation in #22751.  The problem is that two tests still fail with this implementation.  We need to pin down the design spec more explicitly regardless.

core.ops has three functions for defining DataFrame ops: _arith_method_FRAME, _flex_comp_method_FRAME, _comp_method_FRAME.  The first two both call `_align_method_FRAME`, with `_comp_method_FRAME` being the outlier.  This PR just adds that alignment call.

The two tests that currently fail:

```
df = DataFrame(np.arange(6).reshape((3, 2)))
b_r = np.atleast_2d([2, 2])
l = (2, 2, 2)
expected = DataFrame([[False, False], [False, True], [True, True]])
result = df > l   # <-- raises ValueError under this PR because it has the wrong shape
assert_frame_equal(result, expected)

result = df > b_r   # <-- raises ValueError under this PR because it has the wrong shape
assert_frame_equal(result, expected)
```

```
df = pd.DataFrame(np.arange(6).reshape((3, 2)))
with pytest.raises(ValueError):  # <-- doesnt raise
           df == (2, 2)
```

I understand why the behavior tested by the first test makes sense, but don't see the logic behind having `df == (2, 2)` raise (maybe #4576 holds the answer, will look at that more closely)

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -4923,13 +4923,8 @@ def _combine_match_columns(self, other, func, level=None, try_cast=True):
...
     def _combine_const(self, other, func, errors='raise', try_cast=True):
-        if lib.is_scalar(other) or np.ndim(other) == 0:
-            return ops.dispatch_to_series(self, other, func)
-
-        new_data = self._data.eval(func=func, other=other,
-                                   errors=errors,
-                                   try_cast=try_cast)
-        return self._constructor(new_data)
+        assert lib.is_scalar(other) or np.ndim(other) == 0
+        return ops.dispatch_to_series(self, other, func)
```
Comment (jbrockmendel) [inline pandas/core/frame.py]: If/when this is finalized, we would just remove _combine_const altogether.

**Comment by jbrockmendel** [comment]
Related: #22355.

**Comment by jbrockmendel** [comment]
@jorisvandenbossche this is the "pretty" alternative to #22751, needs attention despite not being green.  Any thoughts on desired behavior?

**Comment by jbrockmendel** [comment]
@pandas-dev/pandas-core this now contains a non-trivial API change, should make sure there is consensus before moving forward.

TL;DR: broadcasting/alignment on DataFrame comparison ops behaves differently from arithmetic ops (#22355).  Both behave differently from numpy convention (#22686).  This PR changes the comparison behavior to match the arithmetic behavior, with the numpy part saved for the next round.

Demonstration (based on tests.frame.test_operators test_boolean_comparison in master):
```
df = pd.DataFrame(np.arange(6).reshape((3, 2)))
b = np.array([2, 2])
b_r = np.atleast_2d([2, 2])
b_c = b_r.T
l = (2, 2, 2)
tup = tuple(l)

cmp_expected = pd.DataFrame([[False, False], [False, True], [True, True]])
arith_expected = pd.DataFrame([[2, 3], [4, 5], [6, 7]])

# For the 1-D array `b`, comparison and arithmetic broadcast the same way
result = df > b
tm.assert_frame_equal(result, cmp_expected)
result = df.values > b
tm.assert_numpy_array_equal(result, cmp_expected.values)
result = df + b
tm.assert_frame_equal(result, arith_expected)
result = df.values + b
tm.assert_numpy_array_equal(result, arith_expected.values)

# In the 2-D case, the broadcasting behavior is different
result = df > b_r
tm.assert_frame_equal(result, cmp_expected)
result = df.values > b_r
tm.assert_numpy_array_equal(result, cmp_expected.values)

result = df.values + b_r
tm.assert_numpy_array_equal(result, arith_expected.values)

with pytest.raises(ValueError):  # we DONT want this to raise
    # ValueError: Unable to coerce to DataFrame, shape must be (3, 2): given (1, 2)
    result = df + b_r


# ditto for tuple/list
result = df > tup  # doesn't match numpy or arithmetic behavior
tm.assert_frame_equal(result, cmp_expected)

with pytest.raises(ValueError):
     # numpy behavior
     result = df.values > tup

with pytest.raises(ValueError):
     # numpy behavior
    result = df.values + tup

with pytest.raises(ValueError):
    # ValueError: Unable to coerce to Series, length must be 2: given 3
    result = df + tup
```

**Comment by TomAugspurger** [comment]
Just to make sure I understand, the changes would be

- DataFrame + 2-d ndarray with matching trailing dimension: Previously raised, now broadcasts
- DataFrame + non-ndarray iterable with matching first dimension: Previously broadcast, now raises

Did I miss any?

**Comment by jbrockmendel** [comment]
> DataFrame + 2-d ndarray with matching trailing dimension: Previously raised, now broadcasts

np.ndarray with shape (1, df.shape[1]): previously broadcasts, now raises <-- *see below

> DataFrame + non-ndarray iterable with matching first dimension: Previously broadcast, now raises

list or tuple with len == df.shape[0] != df.shape[1]: previously broadcasts, now raises <-- AFAICT this behavior is not currently tested

list or tuple with len == df.shape[1]: previously raises, now broadcasts


\* In a separate PR I would like to change the behavior when operating with ndarray with shape (nrows, 1) or (1, ncols) so that both of these broadcast identically to numpy (ATM the (nrows, 1) case raises for both arithmetic and comparison ops).  This change would be for both arithmetic and comparison ops (whereas this PR just changes comparison ops to behave like arithmetic ops)

**Comment by jbrockmendel** [comment]
Update: In a different branch I edited the alignment code so that 2d arrays of (nrows, 1) or (1, ncols) both broadcast for arithmetic ops.  This did not cause any test failures.  So AFAICT this particular behavior may be unintentional.

**Comment by jreback** [comment]
you refernced https://github.com/pandas-dev/pandas/issues/4576 which is indeed behavior we patched in the past, but doesn't necessary make sense for a DataFrame, while for a homogenous ndarray does a bit more.

Have to look at your changes in some detail..

**Comment by jbrockmendel** [comment]
> Have to look at your changes in some detail..

Sure.  The succinct version is my favorite: "Make DataFrame comparison ops behave like DataFrame arithmetic ops".  That change is enabled entirely by the 1 line added in `core.ops`.

File: pandas/tests/frame/test_operators.py [hunk 2]
```diff
@@ -804,13 +805,15 @@ def test_boolean_comparison(self):
...
+            df > l
 
-        result = df > tup
-        assert_frame_equal(result, expected)
+        with pytest.raises(ValueError):
+            # wrong shape
+            result = df > tup
 
-        result = df > b_r
+        result = df > b_r  # broadcasts like ndarray (GH#23000)
```
Comment (jbrockmendel) [inline pandas/tests/frame/test_operators.py]: Changed following #23000 (previously in this PR this comparison raised ValueError)

File: pandas/tests/frame/test_operators.py [hunk 3]
```diff
@@ -827,13 +830,13 @@ def test_boolean_comparison(self):
...
+        with pytest.raises(ValueError):
+            result = df == l
 
-        result = df == tup
-        assert_frame_equal(result, expected)
+        with pytest.raises(ValueError):
+            result = df == tup
 
-        result = df == b_r
+        result = df == b_r  # broadcasts like ndarray (GH#23000)
```
Comment (jbrockmendel) [inline pandas/tests/frame/test_operators.py]: Changed following #23000 (previously in this PR this comparison raised ValueError)

File: doc/source/whatsnew/v0.24.0.txt [hunk 4]
```diff
@@ -487,6 +487,71 @@ Previous Behavior:
...
+   Out[5]:
+          0      1
+   0   True   True
+   1  False  False
+   2  False  False
+   In [6]: df + arr[[0], :]
+   ...
+   ValueError: Unable to coerce to DataFrame, shape must be (3, 2): given (1, 2)
+
+   In [7]: df == (1, 2)  # length matches number of columns; comparison used to raise where arithmetic would broadcast
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.txt]: These code blocks are put in `<pre>` divs in the docs, so long lines will force horizontal scrolling / look extra bad on mobile. Better to continue the input with a

```
In [7]:  ...
   ...: # comment ...
   ...: # maybe multiline
```

File: doc/source/whatsnew/v0.24.0.txt [hunk 5]
```diff
@@ -487,6 +487,71 @@ Previous Behavior:
...
+   0  False   True
+   1   True  False
+   2  False  False
+   In [10]: df + (1, 2, 3)
+   ...
+   ValueError: Unable to coerce to Series, length must be 2: given 3
+
+*Current Behavior*:
+
+.. ipython:: python
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.txt]: Will nee d an `:okexcept` here.

Comment (TomAugspurger) [reply doc/source/whatsnew/v0.24.0.txt]: Can also remove the `In []` prefixes. And hopefully the block continues running when there's an exception in an `okexcept`.

**Review by TomAugspurger — COMMENTED** [review]
There's a merge conflict with master now.

File: doc/source/whatsnew/v0.24.0.txt [hunk 6]
```diff
@@ -487,6 +487,71 @@ Previous Behavior:
...
+
+DataFrame Comparison Operations Broadcasting Changes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Previously, the broadcasting behavior of :class:`DataFrame` comparison
+operations (``==``, ``!=``, ...) was inconsistent with the behavior of
+arithmetic operations (``+``, ``-``, ...).  The behavior of the comparison
+operations has been changed to match the arithmetic operations in these cases.
+(:issue:`22880`)
+
+The affected cases are: operating against a 2-dimensional ``np.ndarray`` with
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply doc/source/whatsnew/v0.24.0.txt]: Just pushed addressing your comments.  This one I had to guess what you had in mind.  LMK if I got it wrong.

File: doc/source/whatsnew/v0.24.0.txt [hunk 7]
```diff
@@ -487,6 +487,77 @@ Previous Behavior:
...
+Previously, the broadcasting behavior of :class:`DataFrame` comparison
+operations (``==``, ``!=``, ...) was inconsistent with the behavior of
+arithmetic operations (``+``, ``-``, ...).  The behavior of the comparison
+operations has been changed to match the arithmetic operations in these cases.
+(:issue:`22880`)
+
+The affected cases are:
+  - operating against a 2-dimensional ``np.ndarray`` with either 1 row or 1 column
+  - a list or tuple with the same length matching the number of rows in the :class:`DataFrame`
+  - a list or tuple with thlength matching the number of columns in the :class:`DataFrame`.
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: the length

File: doc/source/whatsnew/v0.24.0.txt [hunk 8]
```diff
@@ -487,6 +487,77 @@ Previous Behavior:
...
+*Current Behavior*:
+
+.. ipython:: python
+   :okexcept:
+
+   arr = np.arange(6).reshape(3, 2)
+   df = pd.DataFrame(arr)
+   df == arr[[0], :]  # raises just like the next arithmetic operation
+   df + arr[[0], :]
+   df == (1, 2)       # broadcasts just like the next arithmetic operation
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: would prefer that the comments precede the lines, also use multiple ipython blocks to make the rendering more clear (if needed).

File: pandas/tests/frame/test_operators.py (hunk 2 above)
Comment (jreback) [reply pandas/tests/frame/test_operators.py]: can you put comments on the previous line

File: pandas/tests/frame/test_operators.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/frame/test_operators.py]: comments on previous

**Review by jreback — CHANGES_REQUESTED** [review]
big issue I have here is what is the performance degredation. .eval dispatched to numexpr, I *think* that this is now handled in dispatch_to_* but not really sure. We *may* need some asvs to test the broadcasting behaviors (e.g. df + 2-d ndarray)

**Comment by jbrockmendel** [comment]
>  .eval dispatched to numexpr, 

I'm pretty sure this is not accurate.

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@jbrockmendel did you do a check of the performance to be sure there is no implication?

**Comment by jbrockmendel** [comment]
> did you do a check of the performance to be sure there is no implication?

There does appear to be a performance hit in the relevant case, but a) there is room for subsequent optimization and b) the bugs this avoids are much more important than the performance  impact.

```
In [3]: arr = np.arange(600).reshape(30, 20)
In [4]: df = pd.DataFrame(arr)
In [5] %timeit arr == arr[[0], :]    # calibrate based on numpy timings
10000 loops, best of 3: 82.9 µs per loop   <-- master
10000 loops, best of 3: 84 µs per loop      <-- PR

In [6]: ser = pd.Series(arr[0, :])
In [7]: l = list(ser)

In [8]: %timeit df == arr[[0], :]
100 loops, best of 3: 4.42 ms per loop   # <-- master
10 loops, best of 3: 87.5 ms per loop    # <-- PR

In [9]: %timeit df == ser
10 loops, best of 3: 83 ms per loop       <-- master
10 loops, best of 3: 86.9 ms per loop    <-- PR

In [10]: %timeit df == l
ValueError: Invalid broadcasting comparison       <-- master
10 loops, best of 3: 82.4 ms per loop    <-- PR

In [11]: df[1] = df[1].astype('f8')
    ...: df[3] = df[3].astype('uint8')
    ...: df[5] = df[5].view('m8[ns]')
    ...: df[7] = df[7].view('M8[ns]')
    ...: df[9] = df[9].astype('f4')
    ...:
In [12]: %timeit df == arr[[0], :]
ValueError: cannot broadcast shape [(30, 15)] with block values [(1, 20)]   <-- master, BUG
TypeError: cannot compare a TimedeltaIndex with type int64   <-- PR, BUG (in TimedeltaIndex.__eq__; just opened #23063)

In [13]: df[5] = df[5].view('i8').astype('i4')
In [14]: %timeit df == arr[[0], :]
ValueError: cannot broadcast shape [(30, 15)] with block values [(1, 20)]  <-- master, BUG
10 loops, best of 3: 86.7 ms per loop  <-- PR

In [14]: %timeit df == arr[0, :]
ValueError: Invalid broadcasting comparison [array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19])] with block values  <-- master
10 loops, best of 3: 80.9 ms per loop  <-- PR

```

**Comment by jbrockmendel** [comment]
gentle ping.  If this (or #23031) goes through the follow-ups will keep me busy in places orthogonal to PeriodArray constructors

File: doc/source/whatsnew/v0.24.0.txt [hunk 9]
```diff
@@ -487,6 +487,88 @@ Previous Behavior:
...
+DataFrame Comparison Operations Broadcasting Changes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Previously, the broadcasting behavior of :class:`DataFrame` comparison
+operations (``==``, ``!=``, ...) was inconsistent with the behavior of
+arithmetic operations (``+``, ``-``, ...).  The behavior of the comparison
+operations has been changed to match the arithmetic operations in these cases.
+(:issue:`22880`)
+
+The affected cases are:
+  - operating against a 2-dimensional ``np.ndarray`` with either 1 row or 1 column
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.24.0.txt]: Can you indicate here what changes for those cases?

**Comment by jorisvandenbossche** [comment]
Quick general feedback: for the cases that worked before but will raise now, I think we should go through a deprecation cycle.

(looking now more in detail on your explanation of the changes above)

**Comment by jorisvandenbossche** [comment]
> In [8]: %timeit df == arr[[0], :]
100 loops, best of 3: 4.42 ms per loop   # <-- master
10 loops, best of 3: 87.5 ms per loop    # <-- PR

How do you see this performance difference? I though this would now raise an error on this PR?

---

So if I understand correctly, some cases that now work, will raise after this PR, but you plan to let them work again in a later PR? 
That makes it a bit difficult to assess what the overall changes for users would be (eg if a deprecation warning would be needed, see my previous post)

(it's not possible to for example first align arithmetic ops with numpy, and then comparison ops with arithmeric/numpy, so you don't have this back and forth?)

---

Something else to consider: for Index we actually deliberately disabled comparing with a length 1 array. See also the recently opened issue https://github.com/pandas-dev/pandas/issues/23078 where we were planning to actually raise instead of broadcast

**Comment by jbrockmendel** [comment]
> How do you see this performance difference?

If you drill down into this, it is driven by poor performance of cmp(DataFrame, Series), which I think can be improved on a subsequent pass.  In the interim I see this as a uncommon corner case and the performance hit as being less important than internal consistency and bugfixes (just noticed this clsoes #20090)

> I though this would now raise an error on this PR?

That was the case when this PR was opened, but since then #22355 changed the behavior of op(DataFrame, ndarray[ndim=2]) to match numpy broadcasting behavior.

> (it's not possible to for example first align arithmetic ops with numpy, and then comparison ops with arithmeric/numpy, so you don't have this back and forth?)

If I understand correctly, this is similar to what I have in mind for optimizing the DataFrame vs Series op.  

> So if I understand correctly, some cases that now work, will raise after this PR, but you plan to let them work again in a later PR?

I think the only case that falls into this category is a list/tuple (_not_ ndarray or Series) with `length(obj) == len(df) != len(df.columns)`.  Those comparisons currently "work" and this PR causes them to raise (matching arithmetic ops behavior).

I would be open to subsequently changing the behavior of both arithmetic+comparisons to do something like:

```
if len(obj) == len(df.columns):
    [behave like arithmetic ops currently do]
elif len(obj) == len(df):
    [behave like comparison ops currently do]
```

but only if this is made consistent across arithmetic/comparison ops.

File: doc/source/whatsnew/v0.24.0.txt (hunk 9 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v0.24.0.txt]: Will update

**Comment by jorisvandenbossche** [comment]
> > I though this would now raise an error on this PR?

> That was the case when this PR was opened, but since then #22355 changed the behavior of op(DataFrame, ndarray[ndim=2]) to match numpy broadcasting behavior.

But then the whatsnew addition you have in this PR is not up to date anymore? As that still mentions that this case will now raise an error

**Comment by jbrockmendel** [comment]
> But then the whatsnew addition you have in this PR is not up to date anymore

Will update.

**Comment by jorisvandenbossche** [comment]
Can you please give a new summary of what this PR does / changes etc, as I am rather lost now. The comment above about "non-trivial API change" is also not relevant any more?

**Comment by jbrockmendel** [comment]
> Can you please give a new summary of what this PR does / changes etc, as I am rather lost now.

Updated above.

Bottom line: This PR makes DataFrame comparison ops broadcasting consistent with DataFrame arithmetic ops.

**Comment by jorisvandenbossche** [comment]
> update Since #22355 was merged,

That's an open issue, not a merged PR?

**Comment by jbrockmendel** [comment]
Updated, #23000

File: doc/source/whatsnew/v0.24.0.txt (hunk 9 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v0.24.0.txt]: Updated; I think other whatsnew comments addressed.  Note that Previous Behavior blocks refer to 0.23, _not_ pre-this-PR

File: doc/source/whatsnew/v0.24.0.txt [hunk 10]
```diff
@@ -487,6 +487,88 @@ Previous Behavior:
...
+*Current Behavior*:
+
+.. ipython:: python
+   :okexcept:
+
+   arr = np.arange(6).reshape(3, 2)
+   df = pd.DataFrame(arr)
+
+.. ipython:: python
+   # comparison and arithmetic both broadcast
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: I'd prefer you make the comments actual sentences here, which is pretty easy when you already have the blocks.

File: pandas/tests/frame/test_arithmetic.py [hunk 11]
```diff
@@ -48,15 +48,19 @@ def test_mixed_comparison(self):
         assert result.all().all()
 
     def test_df_boolean_comparison_error(self):
-        # GH 4576
+        # GH#4576, GH#22880
         # boolean comparisons with a tuple/list give unexpected results
```
Comment (jreback) [inline pandas/tests/frame/test_arithmetic.py]: can you change this comment

File: pandas/tests/frame/test_operators.py [hunk 12]
```diff
@@ -804,12 +805,15 @@ def test_boolean_comparison(self):
...
 
-        result = df > l
-        assert_frame_equal(result, expected)
+        with pytest.raises(ValueError):
+            # wrong shape
+            df > l
 
-        result = df > tup
-        assert_frame_equal(result, expected)
+        with pytest.raises(ValueError):
```
Comment (jreback) [inline pandas/tests/frame/test_operators.py]: can you assert the error messages in these

**Review by jreback — APPROVED** [review]
lgtm. some minor comments. ping on green.

**Comment by jbrockmendel** [comment]
ping

**Comment by jorisvandenbossche** [comment]
@gfyoung sorry, but Brock and me were still actively discussing this. @jreback I think it was clear that was the case, so please don't say "ping on green" in such a case

**Comment by jorisvandenbossche** [comment]
OK, so I simply reverted the merge. That seemed the easiest way to continue the discussion. It's a bit annoying, but @jbrockmendel you will need to open a new PR since this was cannot be re-opened ..
